lnregtest/lib/network_components.py

```diff
@@ -1096,3 +1103,242 @@ def walletbalance(self):
+        # TODO: if this fails, force stop
+
+    def print_rpc_command(self):
+        """Prints the lightning-cli command to use in the shell for testing."""
+        cmd = ' '.join(self.rpc_command)
+        logger.info("%s:", self.name)
+        logger.info(cmd)
+
+    def setup_datadir(self):
+        """Sets up the cln data folder."""
+        pathlib.Path(self.data_dir).mkdir(parents=True, exist_ok=True)
+        config = cln_config_template.format(
+            name=self.name,
+            port=self.lnport
+        )
+
+        with open(self._cln_config_file, 'w') as f:
+            f.write(config)
+
+    def clear_directory(self):
+        """Deletes the lnd data directory of this node."""
+        logger.debug("%s: Cleaning up cln data directory.", self.name)
+        try:
+            shutil.rmtree(self.data_dir)
+        except FileNotFoundError:
+            logger.debug("%s: Directory already clean.", self.name)
+
+    def getinfo(self):
+        returncode, info = self.rpc(['getinfo'])
+        return info
+
+    def getaddress(self, address_type='p2wkh'):
+        returncode, address = self.rpc(['newaddr'])
+        return address['bech32']
+
+    def _connect(self, pubkey, host):
+        logger.info("%s: Connecting to %s", self.name, pubkey)
+        returncode, info = self.rpc(['connect', pubkey, host])
+        return info
+
+    # TODO: somehow unify sync and async apis
+    def _openchannel(self, pubkey, local_sat, remote_sat):
+        logger.info("%s: Open channel to %s", self.name, pubkey)
+
+        command = ['fundchannel', "id="+pubkey, "feerate=normal", "push_msat="+str(remote_sat*1000), "amount="+str(local_sat)]
+        returncode, info = self.rpc(command)
+        if 'txid' in info:
+            info['funding_txid'] = info['txid']
+
+        return info
+
+    async def _a_openchannel(self, pubkey, local_sat, remote_sat):
+        logger.info("%s: Async open channel to %s", self.name, pubkey)
+        return self._openchannel(pubkey, local_sat, remote_sat)
+
+    def connect_and_openchannel(self, pubkey, host, local_sat, remote_sat):
+        self._connect(pubkey, host)
+        # Retry logic since this fails on macs
+        for i in range(0, 100):
+            info = self._openchannel(pubkey, local_sat, remote_sat)
+            if info and 'funding_txid' in info:
+                funding_txid = info['funding_txid']
+                return funding_txid
+            else:
+                time.sleep(1)
+
+    def disconnect(self, pubkey):
+        logger.info("%s: Disconnecting %s.", self.name, pubkey)
+        command = ['disconnect', pubkey]
+        returncode, info = self.rpc(command)
+        return info
+
+    def set_node_pubkey(self):
+        info = self.getinfo()
+        logger.info(
+            "%s: setting node public key to %s",
+            self.name, info['id'])
+        self.pubkey = info['id']
+
+    # Not supported yet
+    def _listchannels(self) -> List[ChannelState]:
```

Comment:
What's the point of this and how does it relate to the method below?

Reply:
Good that you pointed that out. Basically new versions of CLN have `listpeerchannels` command which is much better. For old versions we need to use `listfunds`(which has less data).